astropy/io/fits/convenience.py

```diff
@@ -186,22 +186,34 @@ def getdata(filename, *args, header=None, lower=None, upper=None, view=None,
 
         If the optional keyword ``header`` is set to `True`, this
         function will return a (``data``, ``header``) tuple.
+
+    Raises
+    ------
+    ValueError
+        If HDU has no data.
     """
 
     mode, closed = _get_file_mode(filename)
 
+    ext = kwargs.get('ext')
+    extname = kwargs.get('extname')
+    extver = kwargs.get('extver')
+    no_user_ext = (len(args) == 0 and ext is None and
+                   extname is None and extver is None)
+
     hdulist, extidx = _getext(filename, mode, *args, **kwargs)
     try:
         hdu = hdulist[extidx]
         data = hdu.data
-        if data is None and extidx == 0:
+        if data is None and no_user_ext:
+            # fallback to the first non-primary extension
             try:
                 hdu = hdulist[1]
                 data = hdu.data
             except IndexError:
-                raise IndexError('No data in this HDU.')
+                pass
         if data is None:
-            raise IndexError('No data in this HDU.')
+            raise ValueError('No data in this HDU.')
```

Comment:
Thinking more about that, I'm not sure `ValueError` is more clear than `IndexError`. 

`ValueError`:
> Raised when an operation or function receives an argument that has the right type but an inappropriate value, and the situation is not described by a more precise exception such as IndexError.

`IndexError`:
> Raised when a sequence subscript is out of range.

Since this is in the `getdata` function, the extension raised indicates that there is no data at the specified index. So `ValueError` would be fine but I don't see a problem with `IndexError`, and not changing allows to avoid breaking software that would depend on this.

@dhomeier ?

Reply:
From the definition `IndexError` is the more specific exception, so it should be given preference...
And `hdulist[i].data` not existing should count as an `IndexError` in all but a very strict semantical sense.
The only real difference I can see is that where `hdulist[i].data` does not exist, `hdulist[i+1].data` still may (in fact is the standard situation here for `i=0`), which one would normally not expect for an `IndexError`.
But I think the chance anyone downstream is misled by this is far smaller than the risk of breaking something where currently only the `IndexError` is caught.

Reply:
My interpretation of point 3 in https://github.com/astropy/astropy/issues/11795#issue-902935510 was that the error does not reflect the fact there is missing data within an existing HDU rather than an HDU missing from the HDU list. However, if `IndexError` is still preferred, I'm more than happy to make the change.

Reply:
Yes, that was my original meaning, but I'd agree now that `IndexError` is probably more useful here.